### ISSUE TITLE
A directory which is "excluded" via analysis options should not be searched for other analysis options.

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#56471]
TL;DR What use case is there for excluding a directory from analysis, but desiring to still respect any analysis options files found inside an excluded directory?

This idea primarily comes from a chat @matanlurey and I had. That chat followed a chat thread with @bwilkerson, @keertip, @scheglov, @pq, and @jakemac53.

Matan notes that the Flutter engine repo was changed recently such that all third party code which is vendored in, is now located _in_ the repository, at `third_party/`. It was previously located _outside_ the repository, at `../third_party/`. And it comes with a lot of static analysis reports, errors, warnings, TODOs, lints, etc. When Matan tried to exclude `./third_party` from analysis, via the root analysis options, it didn't work. (This vendored directory is sort of a pub cache, managed by `gclient` rather than `dart pub`.)

We determined that this is sort of working as intended: the analyzer knows about the excluded files, and will not report static errors to an IDE (or `stdout` when invoked via `dart analyze`). But the analyzer still chooses to walk the directory tree in the analysis root, looking for `analysis_options.yaml` files (and I believe `pubspec.yaml` files and `.dart_tool/package_config.json` files), which all inform analyzer about how to analyze files in the tree under an analysis options files (and how to resolve code under a pubspec or package config file). Any such deep analysis options files _do not inherit_ any options from a parent analysis options file (that only happens with an explicit `include:` field). The deep options are fresh, and may not exclude the same files. In the case of the Flutter engine repo, of course none of the analysis options files in `third_party/` exclude their entire contents; it's just vendored code.

I think the team had all agreed on this intended functionality. But Matan and I then tried to figure out why that might be important.

**The question is:** what use case is there for excluding a directory from analysis, but desiring to still respect any analysis options files found inside an excluded directory?

Matan and I couldn't think of any. Certainly, for this case of a directory of vendored code, if it is acting like a pub cache, then you don't want to see any static issues contained (in order to follow our behavior of not analyzing code outside the analysis root, like code in the global pub cache). Even for the general case: if the developer has specified: "exclude directory X: don't analyze inside directory X" and then an analysis options file is found inside directory X (maybe deep), that doesn't mean we should start analyzing files there.

Not respecting analysis options files in excluded directories would be a breaking change, which I think is fine.

**Comment by ThomasDuffinMears** [comment]
EDIT:

```yaml
# workspace_root/analysis_options.yaml

# typical analyzer/linter gubbins

analyzer:
  plugins:
    # ...
  errors:
    # ...
  exclude:
    [
      "**/example/**", # Matches subdirs of a descendant "example" dir eg. workspace_root/packages/my_package/example
      "**/my_package/**", # Temporarily suppressing analyzer in this package while I work on it!
      "**/*.g.dart", # Matches all subdirs that contain .g.dart files.
       "**/*.freezed.dart",
      "**/*.gen.dart",
    ]
```

I think I was _actually_ hitting a wall with how globs work. I was previously pre-pending globstars to the patterns I was looking for:
```sh
      "**/packages/**/example/**" 
      "**/lib/**/*.g.dart"
```

Because I thought I was saying "search every subdir for a `packages` dir, then recurse _that_ to match X file" when globstar handles the recursion for you (unless you overload it and break shit like I did...). TLDR this _seems_ to work for nested packages in a monorepo, but YMMV.

I did realise that perhaps the analyzer _should_ recurse excluded subdirectories, though, as 

>analysis_options _should not override settings implicitly_, they should _inherit settings implicitly_. If I want to override the behavior of analysis in a subdirectory, **then I should use a _new_ yaml file to explicitly establish the rules of the child directories**. If my parent directory excluded files for a subdirectory that I want to **_reinclude_ in analysis, then I should be given an `include` field beneath `analyzer` to map these files.**

Wouldn't be possible if the excluded subdirectories are never parsed. Doing a BFS style search would let you, however, build a graph or some kind of merged yaml to run the analyzer against. Though that final artifact should be surfaced in the tooling or else debugging it as an end-user could become a nightmare.

***

<details><summary>prev comment</summary>
<p>

I think I'm running into this too.

With the move to dart workspaces for monorepos, I wanted to centralize our analysis options - it adds a lot of confusion to devs when you're editing multiple files with the same name and trying to to copy settings over.

The fact that "include" statements are provided in the schemas, you would assume that they work similar to most other languages, including dart, whereby _you're implicitly including the declarations of the imported library_. I don't know _why_ in YAML every implementation seems to ignore this convention, but it makes things *super* cumbersome when X Y Z configuration is inherited but others are not.

analysis_options _should not override settings implicitly_, they should _inherit settings implicitly_. If I want to override the behavior of analysis in a subdirectory, then I should use a _new_ yaml file to explicitly establish the rules of the child directories. If my parent directory excluded files for a subdirectory that I want to _reinclude_ in analysis, then I should be given an `include` field beneath `analyzer` to map these files.

This whole nonsense of overlapping configurations is a nightmare to reason with and makes sensible strategies for project management seem like codesmells.

</p>
</details>
